backend/app/data_products/output_ports/service.py

```diff
@@ -126,21 +161,55 @@ def search_datasets(
         self, query: str, limit: int, user: UserModel
     ) -> Sequence[SearchDatasets]:
         load_options = get_dataset_load_options()
-        stmt = (
-            select(
-                DatasetModel,
-                func.ts_rank_cd(
-                    DatasetModel.search_vector,
-                    func.websearch_to_tsquery("english", query),
-                    32,  # 32 normalizes all results between 0-1
-                ).label("rank"),
+        web_ts = func.websearch_to_tsquery("english", query)
+        prefix_query = self._build_prefix_tsquery(query)
+        # Prepare infix (substring) fallback for cases like 'inica' -> 'Clinical'
+        # Tokenization mirrors _build_prefix_tsquery behavior
+        lowered = query.lower() if query else ""
+        raw_tokens = re.split(r"[\W_]+", lowered) if lowered else []
+        tokens = [t for t in raw_tokens if len(t) >= 2]
+        # Use the highest rank from websearch_to_tsquery and prefix to_tsquery
+        if prefix_query:
+            prefix_ts = func.to_tsquery("english", prefix_query)
+            web_rank_expr = func.ts_rank_cd(
+                DatasetModel.search_vector,
+                web_ts,
+                32,  # normalize between 0-1
             )
-            .options(*load_options)
-            .where(
-                DatasetModel.search_vector.op("@@")(
-                    func.websearch_to_tsquery("english", query)
-                )
+            prefix_rank_expr = func.ts_rank_cd(
+                DatasetModel.search_vector,
+                prefix_ts,
+                32,
             )
+            rank_expr = func.greatest(web_rank_expr, prefix_rank_expr).label("rank")
+            where_clause = DatasetModel.search_vector.op("@@")(
+                web_ts
+            ) | DatasetModel.search_vector.op("@@")(prefix_ts)
+            # Fallback infix matching: require all tokens to appear (substring) in name or description
+            # Apply whenever we have non-trivial tokens to support infix search like 'inica' -> 'Clinical'
+            if tokens:
+                from sqlalchemy import and_, or_
+
+                like_terms = [
+                    or_(
+                        DatasetModel.name.ilike(f"%{t}%"),
+                        DatasetModel.description.ilike(f"%{t}%"),
+                    )
+                    for t in tokens
+                ]
+                where_clause = where_clause | and_(*like_terms)
+        else:
+            rank_expr = func.ts_rank_cd(
+                DatasetModel.search_vector,
+                web_ts,
+                32,  # normalize between 0-1
+            ).label("rank")
+            where_clause = DatasetModel.search_vector.op("@@")(web_ts)
+
+        stmt = (
+            select(DatasetModel, rank_expr)
+            .options(*load_options)
+            .where(where_clause)
```

Comment:
This code changed completely since I last reviewed it and to be honest, the previous version was a lot better to read and simpler, I would prefer that version